src/rtsp/stream_decoding.md

```diff
@@ -0,0 +1,126 @@
+and PPS (Picture Parameter Set). These may be transmitted in-band (together with the video stream, contained inside RTP
+packets) - if that's the case, no problems here. However, they may also be transmitted out-of-band, meaning they're absent
+in the video stream, and have to be delivered in some other way (and that's precisely what most RTSP servers will be doing).
+
+Usually, RTSP servers will transmit stream parameters, together with other useful info, inside a response to the DESCRIBE request.
+They will be encoded using SDP (Session Description Protocol, [RFC 8866](https://www.rfc-editor.org/rfc/rfc6871),
+parameters: [RFC 6871](https://www.rfc-editor.org/rfc/rfc6871)). This means that one needs to parse the parameters from the response,
+then either include them in the RTP stream (which poses another set of problems, more on them below) or pass them on
+externally to the elements further downstream, responsible for decoding the H264 stream.
+
+### Injecting parameters into the stream
+Should you wish to include the parameters in the RTP stream more than once (perhaps adding them before every keyframe,
+so that new peers are able to decode the stream if they joined later on), you immediately run into issues regarding packet numbering. 
+
+Suppose we have the following RTP packets being sent by the server:
+```
+0  1  2  3  4  5  6  7  8  9  10    - sequence number
+      I           I           I     - I if keyframe
+```
+Let's say we have the SPS and PPS (delivered out-of-band) parsed and payloaded into an RTP packet `"P"` we may inject into the stream at will.
+Let's also assume we will drop packets until we get the first keyframe, then inject the parameters before every keyframe in the stream.
+
+If we inject some packets into the stream, we have to change sequence numbers of all following packets. For now, let's say
+we're just going to assign them entirely new numbers, sequentially.
+
+If we receive the packets in sequence, we'll be sending:
+```
+P  2  3  4  5  P  6  7  8  9  P  10    - original sequence numbers
+0  1  2  3  4  5  6  7  8  9  10 11    - new sequence numbers
+```
+
+Suppose, however, that packet 6 arrived out of order, after packets 7 and 8:
+```
+0  1  2  3  4  5  7  8  6  9  10       - packets received by client
+```
+We're just checking if something is a keyframe, so this means we're sending
+```
+P  2  3  4  5  7  8  P  6  9  P  10
+0  1  2  3  4  5  6  7  8  9  10 11    - new sequence numbers
+```
+The issue is obvious, we can't number the packets sequentially in the order we receive them.
+
+OK, then suppose `new = old + offset`, where `offset` - amount of packets injected by us up to that point.
+In sequence:
+```
+P  2  3  4  5  P  6  7  8  9  P  10
+2  3  4  5  6  7  8  9  10 11 12 13    - new sequence numbers
+```
+Seems good, right? Unfortunately, the same pitfall applies here. Consider, once again, the previous order:
+```
+P  2  3  4  5  7  8  P  6  9  P  10
+2  3  4  5  6  8  9  ?? 7  10 11 12
```

Comment:
Actually, wouldn't it work if we remember for each lost packet the number of P included? E.g.

```
P 2 3 5 P 6 8 4 7
```

we have 1xP for 4 and 2xP for 7 so we rewrite as 

```
P 2 3 5 P 6 8 4 7
2 3 4 6 7 8 10 5 9
```

Reply:
Yeah, this works -- provided the lost packets aren't keyframes. If they are, we're left with no space to insert P at:
```
P 2 3 5 P 6  8 4 7  9 11 12  P 10
2 3 4 6 7 8 10 5 9 11 13 14 ?? 12
```
Might be best to state this explicitly, that the keyframes being out-of-order is causing the problem.